### PR TITLE
[clang][NFC] Refactor `clang::Linkage`

### PR DESCRIPTION
**Pull request by Endilll**
This patch introduces a new enumerator `Invalid = 0`, shifting other enumerators by +1. Contrary to how it might sound, this actually affirms status quo of how this enum is stored in `clang::Decl`:
```
  /// If 0, we have not computed the linkage of this declaration.
  /// Otherwise, it is the linkage + 1.
  mutable unsigned CacheValidAndLinkage : 3;
```
This patch makes debuggers to not be mistaken about enumerator stored in this bit-field. It also converts `clang::Linkage` to a scoped enum.

### CODE REVIEW

File: clang/lib/Serialization/ASTWriterDecl.cpp [hunk 1]
```diff
@@ -2214,7 +2214,7 @@ void ASTWriter::WriteDeclAbbrevs() {
   Abv->Add(BitCodeAbbrevOp(0));                       // TSCSpec
   Abv->Add(BitCodeAbbrevOp(0));                       // InitStyle
   Abv->Add(BitCodeAbbrevOp(BitCodeAbbrevOp::Fixed, 1)); // isARCPseudoStrong
-  Abv->Add(BitCodeAbbrevOp(0));                       // Linkage
+  Abv->Add(BitCodeAbbrevOp(1));                         // Linkage
```
Comment (Endilll) [inline clang/lib/Serialization/ASTWriterDecl.cpp:2217]: @ChuanqiXu9 confirmed that touching this is fine (thank you!). If I guess correctly that this value corresponds to `Linkage::None` we have in some default constructors, we should consider changing that default to `Linkage::Invalid`.

`Linkage::None` being default feels like we computed the linkage and arrived at `Linkage::None` even when it's not the case.

CC @AaronBallman

File: clang/include/clang/Basic/Linkage.h [hunk 2]
```diff
@@ -84,22 +88,33 @@ inline bool isUniqueGVALinkage(GVALinkage L) {
 }
 
 inline bool isExternallyVisible(Linkage L) {
-  return L >= VisibleNoLinkage;
+  switch (L) {
+  case Linkage::Invalid:
+    llvm_unreachable("Linkage hasn't been computed!");
```
Comment (Endilll) [inline clang/include/clang/Basic/Linkage.h:93]: @AaronBallman I sprinkled a lot of `llvm_unreachable` in this patch, following an example I've seen in one of the switches on `Linkage`. Let me know if you'd like this to be handled in a different way.

Comment (AaronBallman) [reply clang/include/clang/Basic/Linkage.h:93]: I think this is a good approach.

File: clang/lib/AST/Decl.cpp [hunk 3]
```diff
@@ -1921,7 +1920,20 @@ bool NamedDecl::declarationReplaces(NamedDecl *OldD, bool IsKnownNewer) const {
...
+  switch (getFormalLinkage()) {
+  case Linkage::Invalid:
+    llvm_unreachable("Linkage hasn't been computed!");
+  case Linkage::None:
+    return false;
+  case Linkage::Internal:
+    return true;
+  case Linkage::UniqueExternal:
+  case Linkage::VisibleNone:
+    llvm_unreachable("Non-formal linkage is not allowed here!");
```
Comment (AaronBallman) [inline clang/lib/AST/Decl.cpp:1932]: Can you explain this change a bit? This looks like a functional change?

File: clang/lib/Serialization/ASTWriterDecl.cpp (hunk 1 above)
Comment (AaronBallman) [reply clang/lib/Serialization/ASTWriterDecl.cpp:2217]: Agreed that those should be swapped to `Invalid` instead of `None` (better done as a separate follow-up patch as it isn't an NFC change). However, this is a change to the serialization format so a new PCH consumed by an older Clang will misbehave. Please check whether `VERSION_MAJOR` (in `ASTBitCodes.h`) has been updated for this release and if it hasn't been, increment it by one.

File: clang/tools/libclang/CXIndexDataConsumer.cpp [hunk 4]
```diff
@@ -560,7 +562,7 @@ bool CXIndexDataConsumer::handleDecl(const NamedDecl *D,
 
   if (shouldSuppressRefs())
     markEntityOccurrenceInFile(D, Loc);
-  
+
```
Comment (AaronBallman) [inline clang/tools/libclang/CXIndexDataConsumer.cpp:565]: Every other change in this file is unrelated to the linkage enum and should be backed out (committed separately).

File: clang/lib/Serialization/ASTWriterDecl.cpp (hunk 1 above)
Comment (ChuanqiXu9) [reply clang/lib/Serialization/ASTWriterDecl.cpp:2217]: > If I guess correctly that this value corresponds to Linkage::None we have in some default constructors, we should consider changing that default to Linkage::Invalid.

No. It is not the case. Sorry for didn't explaining things clearly. The value here means the linkage for ParmVarDecl according to line [2185](https://github.com/llvm/llvm-project/pull/71049/files#diff-125f472e690aa3d973bc42aa3c5d580226c5c47661551aca2889f960681aa64dR2185). So it is natural that ParmVarDecl doesn't have a linkage.

Comment (ChuanqiXu9) [inline clang/lib/Serialization/ASTWriterDecl.cpp:2217]:
```suggestion
  Abv->Add(BitCodeAbbrevOp(1));                         // Linkage::None
```

nit

**Review by ChuanqiXu9 — APPROVED** [review]
LGTM. I didn't understand the full context in the private chat.

File: clang/lib/Serialization/ASTWriterDecl.cpp (hunk 1 above)
Comment (AaronBallman) [reply clang/lib/Serialization/ASTWriterDecl.cpp:2217]: Ohhhh! Thanks. None makes far more sense now. Still a change to the serialization values though, so we need to check whether we should do a version bump or not.

Comment (Endilll) [reply clang/lib/Serialization/ASTWriterDecl.cpp:2217]: > Please check whether VERSION_MAJOR (in ASTBitCodes.h) has been updated for this release and if it hasn't been, increment it by one.

This has been bumped after 17 cut-off at least once: https://github.com/llvm/llvm-project/commit/6fb08d8f558a6f28db7835acdb88cab83aea2eb4

Comment (AaronBallman) [reply clang/lib/Serialization/ASTWriterDecl.cpp:2217]: Excellent, thank you! Hmmm, we should probably put a comment there like "this value was last bumped for the Clang NN release" so we don't have to do so much work in the future... (Not as part of these changes, just thinking out loud.)

**Review by AaronBallman — APPROVED** [review]
LGTM aside from the comment nit from @ChuanqiXu9

File: clang/lib/Serialization/ASTWriterDecl.cpp (hunk 1 above)
Comment (Endilll) [reply clang/lib/Serialization/ASTWriterDecl.cpp]: Applied!

File: clang/include/clang/AST/Type.h [hunk 5]
```diff
@@ -1996,7 +1996,7 @@ class alignas(TypeAlignment) Type : public ExtQualsTypeCommonBase {
     TypeBits.Dependence = static_cast<unsigned>(Dependence);
     TypeBits.CacheValid = false;
     TypeBits.CachedLocalOrUnnamed = false;
-    TypeBits.CachedLinkage = NoLinkage;
+    TypeBits.CachedLinkage = llvm::to_underlying(Linkage::Invalid);
```
Comment (Endilll) [inline clang/include/clang/AST/Type.h:1999]: CC @AaronBallman This passes Clang tests with all those `llvm_unreachable` around, so I'm rather confident about this. Note that I removed `NFC` tag from the title.

**Review by AaronBallman — APPROVED** [review]
LGTM but please add NFC back into the patch title so it's more clear why there's no changes to test coverage. Thank you for the cleanup!
